MapleServer2/Types/Trophy.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Generic;
+using MapleServer2.Data.Static;
+using MapleServer2.Packets;
+using MapleServer2.Servers.Game;
+using Maple2Storage.Enums;
+using MapleServer2.Enums;
+using Maple2Storage.Types.Metadata;
+
+namespace MapleServer2.Types
+{
+    public class Trophy
+    {
+        public int Id { get; private set; }
+        public int NextGrade { get; private set; }
+        public int MaxGrade { get; private set; }
+        public long Counter { get; private set; }
+        public long Condition { get; private set; }
+        public bool IsDone { get; private set; }
+        public List<long> Timestamps { get; private set; }
+
+        public Trophy(int trophyId, int grade = 1, int counter = 0, List<long> timestamps = null, bool isDone = false)
+        {
+            Id = trophyId;
+            NextGrade = grade;
+            Counter = counter;
+            Timestamps = timestamps ?? new List<long>();
+            MaxGrade = TrophyMetadataStorage.GetNumGrades(Id);
+            Condition = TrophyMetadataStorage.GetGrade(Id, NextGrade).Condition;
+            IsDone = isDone;
+        }
+
+        public TrophyPacket.GradeStatus GetGradeStatus()
+        {
+            return IsDone ? TrophyPacket.GradeStatus.FinalGrade : TrophyPacket.GradeStatus.NotFinalGrade;
+        }
+
+        public void AddCounter(GameSession session, long amount)
+        {
+            if (IsDone)
+            {
+                return;
+            }
+            Counter += amount;
+
+            if (Counter >= Condition)
+            {
+                ProvideReward(session);
+                NextGrade++;
+                // level up but not completed
+                if (NextGrade <= MaxGrade)
+                {
+                    Condition = TrophyMetadataStorage.GetGrade(Id, NextGrade).Condition;
+                }
+                // level up and completed
+                else
+                {
+                    IsDone = true;
+                    NextGrade--;
+                    string[] cats = TrophyMetadataStorage.GetMetadata(Id).Categories;
+                    foreach (string cat in cats)
```

Comment:
Maybe rename cats to `categories` and cat to `categorie`? Because it doesn't look better simplified like that